File: Java/String/Trie.java





static class Node {
        int cont;
        Node child[];

        public Node() {
            this.cont = 0;
            child = new Node[26];
        }
    }

    static class Trie {
        Node root;

        Trie() {
            root = new Node();
        }

        void insert(String s) {
            Node curr = root;
            for (int i = 0; i < s.length(); ++i) {
                char c = s.charAt(i);
                if (curr.child[c - 'a'] == null) {
                    curr.child[c - 'a'] = new Node();
                }
                curr.child[c - 'a'].cont++;
                curr = curr.child[c - 'a'];
            }
        }


        /* 
        
            if (x.first == 0 && x.second == 0 || len < s.length()) {
                sa.write("NO ESTÁ\n");
            } else {
                sa.write("SI ESTÁ\n");
            }

         */
        pair<Integer, Integer> query(String s) {
            Node curr = root;
            for (int i = 0; i < s.length(); ++i) {
                char c = s.charAt(i);
                if (curr.child[c - 'a'] == null) {
                    return new pair<Integer, Integer>(i, curr.cont);
                }
                curr = curr.child[c - 'a'];
            }
            return new pair<Integer, Integer>(s.length(), curr.cont);
        }
    }

File: cpp/Strings/Trie.cpp

int maxlen;
int cnt;
struct Node {
  int cnt;
  Node* child[26];
};
 
struct Trie {
  Node* root;
  Trie() {
    root = new Node();
  }
 
  void insert(const string &s) {
    Node* cur = root;
    int len = 0;
    forn(i, sz(s)) {
      ++len;
      int c = s[i] - 'a';
      if (cur->child[c] == NULL) {
        cur->child[c] = new Node();
      }
      cur->child[c]->cnt++;
      cur = cur->child[c];
    }
  }

  void dfs(Node* u, int dep){
    if(u->cnt >=3 && dep > maxlen){
      maxlen = dep;
      cnt = u->cnt;
    }
    forn(v, 26)
      if(u->child[v]!=NULL)
        dfs(u->child[v], dep + 1);
  }

  pair<int, int> query(const string &s) {
    Node* cur = root;
    forn (i, sz(s)) {
      int c = s[i] - 'a';
      if (cur->child[c] == NULL) {
        return {i, cur->cnt};
      }
      cur = cur->child[c];
    }
    return {sz(s), cur->cnt};
  }
 
};